spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/convert/QueryMapper.java

```diff
@@ -295,16 +308,13 @@ private Condition mapCondition(CriteriaDefinition criteria, MapSqlParameterSourc
 
 			mappedValue = convertValue(comparator, settableValue.getValue(), propertyField.getTypeHint());
 			sqlType = getTypeHint(mappedValue, actualType.getType(), settableValue);
-		} else if (criteria.getValue() instanceof ValueFunction) {
+		} else if (criteria.getValue() instanceof ValueFunction valueFunction) {
 
-			ValueFunction<Object> valueFunction = (ValueFunction<Object>) criteria.getValue();
-			Object value = valueFunction.apply(getEscaper(comparator));
-
-			mappedValue = convertValue(comparator, value, propertyField.getTypeHint());
```

Comment:
The conversion is entirely gone. Please reinstate conversion by wrapping `ValueFunction` in another function that applies `convertValue`

Reply:
There really never was a conversion. `ValueFunction` is always a `ValueFunction<String>` obtained using `toString` of whatever was passed in as an argument. It also should only be converted to `String` since it is used in `LIKE` expressions and nothing else.

I therefore don't really consider removing the call to the converter a problem. 
Although I also agree, that we should probably implement proper conversion, instead of just calling `toString`. And maybe at the same time at least warn, when `LIKE` is used with a non `String` property.
But I'd rather do that in a separate issue. 
What do you think?

Reply:
One is how it is used. The other is the concept of abstractions. Moving such parts out of sight easily breaks the conceptual flow and then we end up with code that we don't understand. I also think that we have areas in our code that require much more cleanup than this one.  

> And maybe at the same time at least warn, when LIKE is used with a non String property.

I do not agree. One can have an `Enum` property and run a `LIKE` query handing in a `String` value.

Reply:
I introduced conversion of `ValueFunction` in the correct order: first conversion, than escaping, for both JDBC and R2DBC.